packages/grafana-ui/src/components/DropZone/DropZone.tsx

```diff
@@ -0,0 +1,216 @@
+   */
+  onLoad?: (result: string | ArrayBuffer | null) => void;
+}
+
+export interface DropzoneFile {
+  file: File;
+  id: string;
+  error: DOMException | null;
+  progress?: number;
+  abortUpload?: () => void;
+  retryUpload?: () => void;
+}
+
+export function Dropzone({ options, children, readAs, onLoad }: DropzoneProps) {
+  const [files, setFiles] = useState<DropzoneFile[]>([]);
+
+  const setFileProperty = useCallback(
+    (customFile: DropzoneFile, action: (customFileToModify: DropzoneFile) => void) => {
+      setFiles((oldFiles) => {
+        const newFiles = oldFiles.map((f) => {
+          if (f.id === customFile.id) {
+            action(f);
+            return f;
+          }
+          return f;
+        });
+        return newFiles;
+      });
+    },
+    []
+  );
+
+  const onDrop = useCallback(
+    (acceptedFiles: File[], rejectedFiles: FileRejection[], event: DropEvent) => {
+      let customFiles = acceptedFiles.map(mapToCustomFile);
+      if (options?.multiple === false) {
+        setFiles(customFiles);
+      } else {
+        setFiles((oldFiles) => [...oldFiles, ...customFiles]);
+      }
+
+      if (options?.onDrop) {
+        options.onDrop(acceptedFiles, rejectedFiles, event);
+      } else {
+        for (const customFile of customFiles) {
+          const reader = new FileReader();
+
+          const read = () => {
+            if (readAs) {
+              reader[readAs](customFile.file);
+            } else {
+              reader.readAsText(customFile.file);
+            }
+          };
+
+          // Set abort FileReader
+          setFileProperty(customFile, (fileToModify) => {
+            fileToModify.abortUpload = () => {
+              reader.abort();
+            };
+            fileToModify.retryUpload = () => {
+              setFileProperty(customFile, (fileToModify) => {
+                fileToModify.error = null;
+                fileToModify.progress = undefined;
+              });
+              read();
+            };
+          });
+
+          reader.onabort = () => {
+            setFileProperty(customFile, (fileToModify) => {
+              fileToModify.error = new DOMException('Aborted');
+            });
+          };
+
+          reader.onprogress = (event) => {
+            setFileProperty(customFile, (fileToModify) => {
+              fileToModify.progress = event.loaded;
+            });
+          };
+
+          reader.onload = () => {
+            onLoad?.(reader.result);
+          };
+
+          reader.onerror = () => {
+            setFileProperty(customFile, (fileToModify) => {
+              fileToModify.error = reader.error;
+            });
+          };
+
+          read();
+        }
+      }
+    },
+    [onLoad, options, readAs, setFileProperty]
+  );
+
+  const removeFile = (file: DropzoneFile) => {
+    const newFiles = [...files];
+    newFiles.splice(
+      newFiles.findIndex((newFile) => file.id === newFile.id),
+      1
+    );
```

Comment:
```suggestion
    const newFiles = files.filter((f) => file.id !== f.id);
```
I think this is simpler to do with filter, which is also an immutable operation.

Reply:
Absolutely!